### ISSUE TITLE
Python: use environment variable for script name

### ISSUE THREAD

**Issue by FRidh** [NixOS/nixpkgs#24525]
## Issue description

Interpreters like Python do not support `exec -a` for setting the name of a script. For Python we currently inject a line in scripts that hardcodes the path to the original script. This works well in case the script is called and used from its original store path. However, if we wrap it again (e.g. with `python.buildEnv`) and the script calls itself, it will call itself in the original store path, and not via the one at the new wrapped path.

## Proposal

Instead of hardcoding the absolute path we could let it check for an environment variable, e.g. `NIX_SCRIPT_NAME` or `NIX_PYTHON_NAME`. We can then pass this environment variable with `set` with the wrapper.

cc @abbradar 

---

By the way, I've started [writing](https://gist.github.com/FRidh/c3e974e34eac738405af699786b462d5) some things about the current Python infra, what the issues are, what we should support and what direction we could take. Its still very much WIP.

**Comment by abbradar** [comment]
(sorry, I was pretty busy for several months and forgot about this) 

Nice idea! We could also clear this environment variable after using: (pseudo code) 

```
if "NIX_SCRIPT_NAME" in sys.env:
  sys.argv[0] = sys.env["NIX_SCRIPT_NAME"] 
  del sys.env["NIX_SCRIPT_NAME"] 
else:
  # Fallback to old code
```

**Comment by FRidh** [comment]
@abbradar exactly. We can then guarantee that it will call exactly the path we want it to call. 

Of course the issue remains that nested wrappers will each use `set` so we either have to find a way to merge or update wrappers, or, we use `prefix` or `suffix` instead of `set` and pick the last or first item from the list.

**Comment by abbradar** [comment]
Yep, `suffix` is a good idea. We still need to fall back to patched argv0 in case one uses `patchPythonScript` directly (I think we have several cases of this).

I can try to prepare the patch next week if you won't make it first.

**Comment by FRidh** [comment]
@abbradar I think we should right away solve some other issues as well, like https://github.com/NixOS/nixpkgs/issues/24128. Unfortunately my shell scripting skills are far below mediocre so I would appreciate help with that.

I'll open a PR where I add a `sitecustomize.py` file to the `site-packages` folder that listens to `NIX_PYTHONPATH`. That way we also do not need to touch `PYTHONPATH`. We can work from there.

**Comment by mmahut** [comment]
Are there any updates on this issue, please
